src/views/virtualmachines/tree/VirtualMachineTreeView.tsx

```diff
@@ -0,0 +1,113 @@
+import { VirtualMachineModel } from 'src/views/dashboard-extensions/utils';
+
+import { useQueryParamsMethods } from '@kubevirt-utils/components/ListPageFilter/hooks/useQueryParamsMethods';
+import { convertResourceArrayToMap, getResourceUrl } from '@kubevirt-utils/resources/shared';
+import { FilterValue, useActiveNamespace } from '@openshift-console/dynamic-plugin-sdk';
+import { Grid, GridItem, TreeView, TreeViewDataItem } from '@patternfly/react-core';
+import { TEXT_FILTER_LABELS_ID } from '@virtualmachines/list/hooks/constants';
+
+import { useSyncClicksEffects } from './hooks/useSyncClicksEffects';
+import { useTreeViewData } from './hooks/useTreeViewData';
+import {
+  FOLDER_SELECTOR_PREFIX,
+  PROJECT_SELECTOR_PREFIX,
+  VM_FOLDER_LABEL,
+} from './utils/constants';
+import {
+  createTreeViewData,
+  selectedTreeItem,
+  setSelectedTreeItem,
+  treeDataMap,
+} from './utils/utils';
+
+import './VirtualMachineTreeView.scss';
+
+type VirtualMachineTreeViewProps = {
+  onFilterChange?: (type: string, value: FilterValue) => void;
+};
+
+const VirtualMachineTreeView: FC<VirtualMachineTreeViewProps> = ({ children, onFilterChange }) => {
+  const navigate = useNavigate();
+  const location = useLocation();
+  const { setOrRemoveQueryArgument } = useQueryParamsMethods();
+  const pathname = location.pathname;
+  const [activeNamespace, setActiveNamespace] = useActiveNamespace();
+
+  const { isAdmin, loaded, loadError, projectNames, vms } = useTreeViewData();
+
+  const treeData = useMemo(() => {
+    const [data, dataMap] = createTreeViewData(
+      projectNames,
+      vms,
+      activeNamespace,
+      isAdmin,
+      pathname,
+    );
+    treeDataMap.value = dataMap;
+    return data;
+  }, [projectNames, vms, activeNamespace, isAdmin, pathname]);
+
+  const vmsMapper = useMemo(() => convertResourceArrayToMap(vms, true), [vms]);
+
+  useSyncClicksEffects(activeNamespace, loaded, location);
+
+  if (loadError) return <>{children}</>;
+
+  const onSelect = (_event: MouseEvent, treeViewItem: TreeViewDataItem) => {
+    setSelectedTreeItem(treeViewItem);
+    onFilterChange?.(TEXT_FILTER_LABELS_ID, null);
+
+    const treeItemName = treeViewItem.name as string;
+    if (treeViewItem.id.startsWith(FOLDER_SELECTOR_PREFIX)) {
+      const [_, folderNamespace] = treeViewItem.id.split('/');
+      navigate(
+        getResourceUrl({
+          activeNamespace: folderNamespace,
+          model: VirtualMachineModel,
+        }),
+      );
+      setOrRemoveQueryArgument(TEXT_FILTER_LABELS_ID, `${VM_FOLDER_LABEL}=${treeItemName}`);
+      return onFilterChange?.(TEXT_FILTER_LABELS_ID, {
+        all: [`${VM_FOLDER_LABEL}=${treeItemName}`],
+      });
+    }
+
+    if (treeViewItem.id.startsWith(PROJECT_SELECTOR_PREFIX)) {
+      setActiveNamespace(treeItemName);
+      return navigate(
+        getResourceUrl({
+          activeNamespace: treeItemName,
+          model: VirtualMachineModel,
+        }),
+      );
```

Comment:
seems redundant those 2 actions, as change in navigate is changing URL , so activeNampace should be change also.. please check